packages/plugin-exec/sources/ExecResolver.ts

```diff
@@ -35,13 +35,24 @@ export class ExecResolver implements Resolver {
     return [];
   }
 
-  async getCandidates(descriptor: Descriptor, dependencies: unknown, opts: ResolveOptions) {
-    let path = descriptor.range;
+  async getCandidates(descriptor: Descriptor, dependencies: Map<DescriptorHash, Package>, opts: ResolveOptions) {
+    if (!opts.fetchOptions)
+      throw new Error(`Assertion failed: This resolver cannot be used unless a fetcher is configured`);
+
+    const {path, parentLocator} = execUtils.parseSpec(descriptor.range);
 
-    if (path.startsWith(PROTOCOL))
-      path = path.slice(PROTOCOL.length);
+    const generatorPath = await execUtils.getGeneratorPath(structUtils.makeRange({
+      protocol: PROTOCOL,
+      source: path,
+      selector: path,
+      params: {
+        // The Descriptor should already be bound
+        locator: structUtils.stringifyLocator(parentLocator!),
+      },
+    }), PROTOCOL, opts.fetchOptions);
+    const generatorHash = (await hashUtils.checksumFile(generatorPath)).slice(0, 6);
```

Comment:
This doesn't work in some cases - the path to the generator only makes sense within the FakeFS returned by the `fetch` method. You shouldn't access it as if it was available on the local filesystem.

Take a look at the `PatchResolver`: first we [load the patchfile content](https://github.com/yarnpkg/berry/blob/master/packages/plugin-patch/sources/PatchResolver.ts#L53), and then we [hash it](https://github.com/yarnpkg/berry/blob/master/packages/plugin-patch/sources/PatchResolver.ts#L59).

Reply:
API-wise, I think mimicking the `patchUtils` function would make sense: can you change `getGeneratorPath` into something like `loadGeneratorFile`? It would be more consistent, and in line with the idea of "don't return a path, but rather return the content" 🙂